test/gtest/core/test_multiproc.cc

```diff
@@ -0,0 +1,203 @@
+    EXPECT_EQ(UCC_OK, ucc_init(&lib_params, lib_config, &lib_h));
+    ucc_lib_config_release(lib_config);
+}
+
+void test_process::init_ctx()
+{
+    EXPECT_EQ(UCC_OK, ucc_context_config_read(lib_h, NULL, &ctx_config));
+    ctx_params.mask     = UCC_CONTEXT_PARAM_FIELD_TYPE;
+    ctx_params.ctx_type = UCC_CONTEXT_EXCLUSIVE;
+    EXPECT_EQ(UCC_OK,
+              ucc_context_create(lib_h, &ctx_params, ctx_config, &ctx_h));
+}
+
+void test_process::fini_lib()
+{
+    EXPECT_EQ(UCC_OK, ucc_finalize(lib_h));
+}
+
+void test_process::fini_ctx()
+{
+    EXPECT_EQ(UCC_OK, ucc_context_destroy(ctx_h));
+}
+
+test_multiproc::test_multiproc(int _n_procs) : n_procs(_n_procs)
+{
+    ag.resize(n_procs);
+    for (int i = 0; i < n_procs; i++) {
+        auto p = new test_process();
+        p->init_lib();
+        p->init_ctx();
+        procs.push_back(p);
+        ag[i].phase = AG_INIT;
+    }
+    copy_complete_count = 0;
+}
+
+test_multiproc::~test_multiproc()
+{
+    for (auto p : procs) {
+        p->fini_ctx();
+        p->fini_lib();
+        delete p;
+    }
+}
+
+ucc_status_t allgather(void *src_buf, void *recv_buf, size_t size,
+                       void *coll_info, void **request)
+{
+    test_multiproc::allgather_coll_info_t *ci =
+        (test_multiproc::allgather_coll_info_t *)coll_info;
+    int my_rank                 = ci->my_rank;
+    ci->self->ag[my_rank].sbuf  = src_buf;
+    ci->self->ag[my_rank].rbuf  = recv_buf;
+    ci->self->ag[my_rank].len   = size;
+    ci->self->ag[my_rank].phase = test_multiproc::AG_READY;
+    *request                    = (void *)ci;
+    return UCC_OK;
+}
+
+ucc_status_t req_test(void *request)
+{
+    test_multiproc::allgather_coll_info_t *ci =
+        (test_multiproc::allgather_coll_info_t *)request;
+    int n_procs = ci->self->n_procs;
+    switch (ci->self->ag[ci->my_rank].phase) {
+    case test_multiproc::AG_READY:
+        for (int i = 0; i < n_procs; i++) {
+            if (ci->self->ag[i].phase == test_multiproc::AG_INIT) {
+                return UCC_INPROGRESS;
+            }
+        }
+        for (int i = 0; i < n_procs; i++) {
+            memcpy((void *)((ptrdiff_t)ci->self->ag[ci->my_rank].rbuf +
+                            i * ci->self->ag[i].len),
+                   ci->self->ag[i].sbuf, ci->self->ag[i].len);
+        }
+        ci->self->ag[ci->my_rank].phase = test_multiproc::AG_COPY_DONE;
+        ;
+        ci->self->copy_complete_count++;
+        break;
+    case test_multiproc::AG_COPY_DONE:
+        if (ci->my_rank == 0 && ci->self->copy_complete_count == n_procs) {
+            for (int i = 0; i < n_procs; i++) {
+                ci->self->ag[i].phase = test_multiproc::AG_COMPLETE;
+            }
+            ci->self->copy_complete_count = 0;
+        }
+        break;
+    case test_multiproc::AG_COMPLETE:
+        return UCC_OK;
+    default:
+        break;
+    }
+    return UCC_INPROGRESS;
+}
+
+ucc_status_t req_free(void *request)
+{
```

Comment:
Don't we need to reset AG state to AG_INIT?

Reply:
great catch 👍